keras/callbacks.py

```diff
@@ -76,48 +82,76 @@ def on_epoch_end(self, epoch, logs=None):
 
-    def on_batch_begin(self, batch, logs=None):
+    def _batch_callbacks(self, batch, logs,
+                         callback_method,
+                         callback_duration_history):
+        logs = logs or {}
+        t_before_callbacks = time.time()
+        for callback in self.callbacks:
+            getattr(callback, callback_method)(batch, logs)
+        history = getattr(self, callback_duration_history)
+        history.append(time.time() - t_before_callbacks)
+        delta_t_median = np.median(history)
+        if (self._delta_t_batch > 0. and
+                delta_t_median > 0.95 * self._delta_t_batch and
+                delta_t_median > 0.1):
+            warnings.warn('Method %s() is slow compared '
+                          'to the batch update (%f). Check your callbacks.'
+                          % (callback_method, delta_t_median))
+
+    def _batch_begin_callbacks(self, *args):
         """Called right before processing a batch.
 
         # Arguments
             batch: integer, index of batch within the current epoch.
             logs: dictionary of logs.
         """
-        logs = logs or {}
-        t_before_callbacks = time.time()
-        for callback in self.callbacks:
-            callback.on_batch_begin(batch, logs)
-        self._delta_ts_batch_begin.append(time.time() - t_before_callbacks)
-        delta_t_median = np.median(self._delta_ts_batch_begin)
-        if (self._delta_t_batch > 0. and
-           delta_t_median > 0.95 * self._delta_t_batch and
-           delta_t_median > 0.1):
-            warnings.warn('Method on_batch_begin() is slow compared '
-                          'to the batch update (%f). Check your callbacks.'
-                          % delta_t_median)
+        self._batch_callbacks(*args)
+
         self._t_enter_batch = time.time()
 
-    def on_batch_end(self, batch, logs=None):
+    def _batch_end_callbacks(self, *args):
         """Called at the end of a batch.
 
         # Arguments
             batch: integer, index of batch within the current epoch.
             logs: dictionary of logs.
         """
-        logs = logs or {}
         if not hasattr(self, '_t_enter_batch'):
             self._t_enter_batch = time.time()
         self._delta_t_batch = time.time() - self._t_enter_batch
-        t_before_callbacks = time.time()
-        for callback in self.callbacks:
-            callback.on_batch_end(batch, logs)
-        self._delta_ts_batch_end.append(time.time() - t_before_callbacks)
-        delta_t_median = np.median(self._delta_ts_batch_end)
-        if (self._delta_t_batch > 0. and
-           (delta_t_median > 0.95 * self._delta_t_batch and delta_t_median > 0.1)):
-            warnings.warn('Method on_batch_end() is slow compared '
-                          'to the batch update (%f). Check your callbacks.'
-                          % delta_t_median)
+
+        self._batch_callbacks(*args)
+
+    def on_fit_batch_begin(self, batch, logs=None):
```

Comment:
The new `fetches` functionality of the `Function` class can be used to get access to extra tensors that are not defined as a model output. This can be useful when training (e.g. to get access to predictions: #8468 - I also used it with `fit_generator` myself) and predicting (e.g. to get the output of an intermediate layer) too.
Callbacks for evaluation methods were requested in #4185 and #9484.

Keeping the method names for the training case might be a cause of confusion from the moment similarly named methods are introduced.
Opting for clear names for the train methods and introducing them at the same time the new hooks are introduced seemed like the best API for the future.

A few things can help preventing confusion among current users:
- adding documentation to the `Callback` class
- switching built-in callbacks to the new method names (with the old methods calling the new ones so extensions of these callbacks keep working)
- adding warnings (actively deprecating the old methods)

I still saw two options for naming the train and validation methods:
- `fit` or `train`
- `evaluate` or `test`
Since `on_train_begin` exists, `on_{train|test|predict}_batch_{begin|end}` might be consistent, also matching the batch methods of `Model`, but the loop and generator methods use `fit` and `evaluate`. What do you think would be the best choice?

Reply:
Hmm, ok makes sense. If we went the train/test way, do you think we could combine `on_evaluate_batch_{begin|end}` and `on_predict_batch_{begin|end}` into `on_test_batch_{begin|end}`? This might make callbacks more reusable across `evaluate` and `predict`, if the behavior they need to implement isn't different. Or are the use cases pretty separate?

Also, can we add methods that mark the start and end of each validation epoch inside a fit loop? I've been working on a change in tf.keras that will require the callback being aware of when fit switches from training to validation

Reply:
The output of validation and prediction is different, so the hook usage will be different too because of other tensor/output access and what it is done with that data.
If the functionality is the same, both hooks can always call a shared function.

Would `on_evaluate_{begin|end}` (see the other comment) be enough for what you want to do in tf.keras?

Reply:
Yes, that is a perfect summary of the Callback class!

Reply:
Sounds good to me, @hermansje if that sounds good to you as well can you update the names to match that convention and then I can approve?

Reply:
Thank you @fchollet, I will update the PR!

Calling `on_batch_{begin | end}` for all modes could break or change behaviour of old user callbacks.
Is this something for the changelog and versioning, or should I do something codewise?

Reply:
@hermansje it may need to be that in `fit` methods `on_batch_{begin|end}` is only called during the training loop and not the validation loops, otherwise yes I think user Callbacks would be broken and some of the built-in Callbacks would need refactoring as well

Reply:
@hermansje Let's have `on_batch_{ begin | end }` be aliases for `on_train_batch_{begin | end}` like in your original code